### PR TITLE
Auth Controller

### PR DESCRIPTION
**Pull request by y-abs**
## What does this PR do ?
Add the documentation for auth controller

Add the documentation for core classes User/Profile/Role/Response/Error-Response

### Other changes

Change `validateMyCredentials` signature, make it consistent with others sdk
Fix profile.getRoles method

### CODE REVIEW

File: .ci/doc/templates/default.tpl.dart [hunk 1]
```diff
@@ -1,6 +1,9 @@
 import 'dart:io';
 
 import 'package:kuzzle/kuzzle.dart';
+import 'package:kuzzle/src/kuzzle/profile.dart';
+import 'package:kuzzle/src/kuzzle/role.dart';
+import 'package:kuzzle/src/kuzzle/user.dart';
```
Comment (jenow) [inline .ci/doc/templates/default.tpl.dart]: These 3 files should be imported in `kuzzle/kuzzle.dart` so no need to import them here. If it's not the case you should import them there.

File: doc/2/controllers/auth/check-token/snippets/check-token.dart [hunk 2]
```diff
@@ -0,0 +1,14 @@
+final jwt = await kuzzle
+  .auth
+  .login(
+    'local',
+    {
+      'username': 'foo',
+      'password': 'bar'
+    },
+    expiresIn:''
```
Comment (jenow) [inline doc/2/controllers/auth/check-token/snippets/check-token.dart]:
```suggestion
    expiresIn: ''
```

File: doc/2/controllers/auth/create-my-credentials/index.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+Future<Map<String, dynamic>> createMyCredentials(
+    String strategy, {
+    Map<String, dynamic> credentials,
+  })
+```
+
+| Argument      | Type               | Description                          |
+|---------------|--------------------|--------------------------------------|
+| `strategy`    | <pre>String</pre>  | Strategy to use                      |
+| `credentials` | <pre>Map<String, dynamic><String, Object></re> | Map representing the credentials |
```
Comment (jenow) [inline doc/2/controllers/auth/create-my-credentials/index.md]:
```suggestion
| `credentials` | <pre>Map<String, dynamic></pre> | Map representing the credentials |
```

File: doc/2/controllers/auth/create-my-credentials/snippets/create-my-credentials.dart [hunk 4]
```diff
@@ -0,0 +1,20 @@
+await kuzzle
+  .auth
+  .login(
+    'local',
+    {
+      'username': 'foo',
+      'password': 'bar'
+    },
+    expiresIn:''
```
Comment (jenow) [inline doc/2/controllers/auth/create-my-credentials/snippets/create-my-credentials.dart]:
```suggestion
    expiresIn: ''
```

File: doc/2/controllers/auth/credentials-exist/snippets/credentials-exist.dart [hunk 5]
```diff
@@ -0,0 +1,20 @@
+final result = await kuzzle
+  .auth
+  .login(
+    'local',
+    {
+      'username': 'foo',
+      'password': 'bar'
+    },
+    expiresIn:''
```
Comment (jenow) [inline doc/2/controllers/auth/credentials-exist/snippets/credentials-exist.dart]:
```suggestion
    expiresIn: ''
```

File: doc/2/controllers/auth/credentials-exist/snippets/credentials-exist.dart [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  );
+
+  final exists = await kuzzle
+  .auth
+  .credentialsExist(
+    'local'
+  );
+
+  if (exists) {
+    print("Credentials exists for local strategy");
```
Comment (jenow) [inline doc/2/controllers/auth/credentials-exist/snippets/credentials-exist.dart]:
```suggestion
    print('Credentials exists for local strategy');
```

File: doc/2/controllers/auth/delete-my-credentials/snippets/delete-my-credentials.dart [hunk 7]
```diff
@@ -0,0 +1,14 @@
+await kuzzle
+  .auth
+  .login(
+    'local',
+    {
+      'username': 'foo',
+      'password': 'bar'
+    },
+    expiresIn:''
```
Comment (jenow) [inline doc/2/controllers/auth/delete-my-credentials/snippets/delete-my-credentials.dart]:
```suggestion
    expiresIn: ''
```

File: doc/2/controllers/auth/get-current-user/index.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
+---
+code: true
+type: page
+title: GetCurrentUser
+description: Returns the profile object for the user linked to the `JSON Web Token`.
+---
+
+# GetCurrentUserAsync
```
Comment (jenow) [inline doc/2/controllers/auth/get-current-user/index.md]:
```suggestion
# GetCurrentUser
```

File: doc/2/controllers/auth/check-token/index.md [hunk 9]
```diff
@@ -0,0 +1,34 @@
+---
+code: true
+type: page
+title: CheckToken
```
Comment (jenow) [inline doc/2/controllers/auth/check-token/index.md]:
```suggestion
title: checkToken
```

File: doc/2/controllers/auth/create-my-credentials/index.md [hunk 10]
```diff
@@ -0,0 +1,32 @@
+---
+code: true
+type: page
+title: CreateMyCredentials
```
Comment (jenow) [inline doc/2/controllers/auth/create-my-credentials/index.md]:
```suggestion
title: createMyCredentials
```

File: doc/2/controllers/auth/credentials-exist/index.md [hunk 11]
```diff
@@ -0,0 +1,28 @@
+---
+code: true
+type: page
+title: CredentialsExist
```
Comment (jenow) [inline doc/2/controllers/auth/credentials-exist/index.md]:
```suggestion
title: credentialsExist
```

File: doc/2/controllers/auth/delete-my-credentials/index.md [hunk 12]
```diff
@@ -0,0 +1,24 @@
+---
+code: true
+type: page
+title: DeleteMyCredentials
```
Comment (jenow) [inline doc/2/controllers/auth/delete-my-credentials/index.md]:
```suggestion
title: deleteMyCredentials
```

File: doc/2/controllers/auth/get-current-user/index.md [hunk 13]
```diff
@@ -0,0 +1,24 @@
+---
+code: true
+type: page
+title: GetCurrentUser
```
Comment (jenow) [inline doc/2/controllers/auth/get-current-user/index.md]:
```suggestion
title: getCurrentUser
```

File: doc/2/controllers/auth/get-current-user/snippets/get-current-user.dart [hunk 14]
```diff
@@ -0,0 +1,14 @@
+await kuzzle
+  .auth
+  .login(
+    'local',
+    {
+      'username': 'foo',
+      'password': 'bar'
+    },
+    expiresIn:''
```
Comment (jenow) [inline doc/2/controllers/auth/get-current-user/snippets/get-current-user.dart]:
```suggestion
    expiresIn: ''
```

File: doc/2/controllers/auth/get-my-credentials/index.md [hunk 15]
```diff
@@ -0,0 +1,28 @@
+---
+code: true
+type: page
+title: GetMyCredentials
```
Comment (jenow) [inline doc/2/controllers/auth/get-my-credentials/index.md]:
```suggestion
title: getMyCredentials
```

File: doc/2/controllers/auth/get-my-credentials/snippets/get-my-credentials.dart [hunk 16]
```diff
@@ -0,0 +1,14 @@
+await kuzzle
+  .auth
+  .login(
+    'local',
+    {
+      'username': 'foo',
+      'password': 'bar'
+    },
+    expiresIn:''
```
Comment (jenow) [inline doc/2/controllers/auth/get-my-credentials/snippets/get-my-credentials.dart]:
```suggestion
    expiresIn: ''
```

File: doc/2/controllers/auth/get-my-rights/index.md [hunk 17]
```diff
@@ -0,0 +1,24 @@
+---
+code: true
+type: page
+title: GetMyRights
```
Comment (jenow) [inline doc/2/controllers/auth/get-my-rights/index.md]:
```suggestion
title: getMyRights
```

File: doc/2/controllers/auth/get-my-rights/snippets/get-my-rights.dart [hunk 18]
```diff
@@ -0,0 +1,14 @@
+await kuzzle
+  .auth
+  .login(
+    'local',
+    {
+      'username': 'foo',
+      'password': 'bar'
+    },
+    expiresIn:''
```
Comment (jenow) [inline doc/2/controllers/auth/get-my-rights/snippets/get-my-rights.dart]:
```suggestion
    expiresIn: ''
```

File: doc/2/controllers/auth/get-strategies/index.md [hunk 19]
```diff
@@ -0,0 +1,24 @@
+---
+code: true
+type: page
+title: GetStrategies
```
Comment (jenow) [inline doc/2/controllers/auth/get-strategies/index.md]:
```suggestion
title: getStrategies
```

File: doc/2/controllers/auth/get-strategies/snippets/get-strategies.dart [hunk 20]
```diff
@@ -0,0 +1,14 @@
+await kuzzle
+  .auth
+  .login(
+    'local',
+    {
+      'username': 'foo',
+      'password': 'bar'
+    },
+    expiresIn:''
```
Comment (jenow) [inline doc/2/controllers/auth/get-strategies/snippets/get-strategies.dart]:
```suggestion
    expiresIn: ''
```

File: doc/2/controllers/auth/login/index.md [hunk 21]
```diff
@@ -0,0 +1,59 @@
+---
+code: true
+type: page
+title: Login
```
Comment (jenow) [inline doc/2/controllers/auth/login/index.md]:
```suggestion
title: login
```

File: doc/2/controllers/auth/login/snippets/login.dart [hunk 22]
```diff
@@ -0,0 +1,10 @@
+final result = await kuzzle
+  .auth
+  .login(
+    'local',
+    {
+      'username': 'foo',
+      'password': 'bar'
+    },
+    expiresIn:''
```
Comment (jenow) [inline doc/2/controllers/auth/login/snippets/login.dart]:
```suggestion
    expiresIn: ''
```

File: doc/2/controllers/auth/logout/index.md [hunk 23]
```diff
@@ -0,0 +1,24 @@
+---
+code: true
+type: page
+title: Logout
```
Comment (jenow) [inline doc/2/controllers/auth/logout/index.md]:
```suggestion
title: logout
```

File: doc/2/controllers/auth/logout/snippets/logout.dart [hunk 24]
```diff
@@ -0,0 +1,14 @@
+await kuzzle
+  .auth
+  .login(
+    'local',
+    {
+      'username': 'foo',
+      'password': 'bar'
+    },
+    expiresIn:''
```
Comment (jenow) [inline doc/2/controllers/auth/logout/snippets/logout.dart]:
```suggestion
    expiresIn: ''
```

File: doc/2/controllers/auth/update-my-credentials/index.md [hunk 25]
```diff
@@ -0,0 +1,35 @@
+---
+code: true
+type: page
+title: UpdateMyCredentials
```
Comment (jenow) [inline doc/2/controllers/auth/update-my-credentials/index.md]:
```suggestion
title: updateMyCredentials
```

File: doc/2/controllers/auth/update-my-credentials/snippets/update-my-credentials.dart [hunk 26]
```diff
@@ -0,0 +1,20 @@
+await kuzzle
+  .auth
+  .login(
+    'local',
+    {
+      'username': 'foo',
+      'password': 'bar'
+    },
+    expiresIn:''
```
Comment (jenow) [inline doc/2/controllers/auth/update-my-credentials/snippets/update-my-credentials.dart]:
```suggestion
    expiresIn: ''
```

File: doc/2/controllers/auth/update-self/index.md [hunk 27]
```diff
@@ -0,0 +1,28 @@
+---
+code: true
+type: page
+title: UpdateSelf
```
Comment (jenow) [inline doc/2/controllers/auth/update-self/index.md]:
```suggestion
title: updateSelf
```

File: doc/2/controllers/auth/update-self/snippets/update-self.dart [hunk 28]
```diff
@@ -0,0 +1,18 @@
+await kuzzle
+  .auth
+  .login(
+    'local',
+    {
+      'username': 'foo',
+      'password': 'bar'
+    },
+    expiresIn:''
```
Comment (jenow) [inline doc/2/controllers/auth/update-self/snippets/update-self.dart]:
```suggestion
    expiresIn: ''
```

File: doc/2/controllers/auth/validate-my-credentials/snippets/validate-my-credentials.dart [hunk 29]
```diff
@@ -0,0 +1,20 @@
+await kuzzle
+  .auth
+  .login(
+    'local',
+    {
+      'username': 'foo',
+      'password': 'bar'
+    },
+    expiresIn:''
```
Comment (jenow) [inline doc/2/controllers/auth/validate-my-credentials/snippets/validate-my-credentials.dart]:
```suggestion
    expiresIn: ''
```

File: doc/2/core-classes/profile/getRoles/snippets/get-roles.dart [hunk 30]
```diff
@@ -0,0 +1,17 @@
+await kuzzle
+  .auth
+  .login(
+    'local',
+    {
+      'username': 'foo',
+      'password': 'bar'
+    },
+    expiresIn:''
```
Comment (jenow) [inline doc/2/core-classes/profile/getRoles/snippets/get-roles.dart]:
```suggestion
    expiresIn: ''
```

File: doc/2/core-classes/response/introduction/index.md [hunk 31]
```diff
@@ -0,0 +1,30 @@
...
+| `index` | <pre>String</pre> | Impacted index |
+| `protocol` | <pre>String</pre> | Network protocol at the origin of the real-time notification |
+| `requestId` | <pre>String</pre> | Request unique identifier |
+| `result` | <pre>Object</pre> | Response payload (depends on the executed API action) |
+| `room` | <pre>String</pre> | Room identifier (realtime only) |
+| `scope` | <pre>String</pre> | Document scope ("in" or "out", realtime only) |
+| `state` | <pre>String</pre> | Document state (realtime only) |
+| `status` | <pre>int</pre> | Response status, following HTTP status codes |
+| `timestamp` | <pre>Long</pre> | Notification timestamp (UTC) |
+| `volatile` | <pre>ConcurrentHashMap<String, Object></pre> | Volatile data |
```
Comment (jenow) [inline doc/2/core-classes/response/introduction/index.md]:
```suggestion
| `volatile` | <pre>Map<String, dynamic></pre> | Volatile data |
```

File: doc/2/core-classes/user/getProfiles/snippets/get-profiles.dart [hunk 32]
```diff
@@ -0,0 +1,16 @@
+await kuzzle
+  .auth
+  .login(
+    'local',
+    {
+      'username': 'foo',
+      'password': 'bar'
+    },
+    expiresIn:''
```
Comment (jenow) [inline doc/2/core-classes/user/getProfiles/snippets/get-profiles.dart]:
```suggestion
    expiresIn: ''
```

**Review by jenow — CHANGES_REQUESTED** [review]

**Review by jenow — APPROVED** [review]

File: lib/src/controllers/auth.dart [hunk 33]
```diff
@@ -67,7 +67,7 @@ class AuthController extends KuzzleController {
       action: 'getCurrentUser',
     ));
 
-    return KuzzleUser.fromKuzzleResponse(kuzzle, response);
+  return KuzzleUser.fromKuzzleResponse(kuzzle, response);
```
Comment (alexandrebouthinon) [inline lib/src/controllers/auth.dart]:
```suggestion
    return KuzzleUser.fromKuzzleResponse(kuzzle, response);
```

**Review by alexandrebouthinon — APPROVED** [review]
I spotted a little indentation error make it pass the CI and you're done :+1:

**Pull request merged**